lib/IDE/Refactoring.cpp

```diff
@@ -7109,42 +7199,23 @@ class AsyncConverter : private SourceEntityWalker {
                                  const AsyncHandlerParamDesc &HandlerDesc,
                                  const ClosureExpr *Callback,
                                  PtrArrayRef<Expr *> ArgList) {
-    ArrayRef<const ParamDecl *> CallbackParams =
-        Callback->getParameters()->getArray();
-    auto CallbackBody = Callback->getBody();
-    if (HandlerDesc.params().size() != CallbackParams.size()) {
+    if (HandlerDesc.params().size() != Callback->getParameters()->size()) {
       DiagEngine.diagnose(CE->getStartLoc(), diag::mismatched_callback_args);
       return;
     }
-
-    // Note that the `ErrParam` may be a Result (in which case it's also the
-    // only element in `SuccessParams`)
-    ArrayRef<const ParamDecl *> SuccessParams = CallbackParams;
-    const ParamDecl *ErrParam = nullptr;
-    if (HandlerDesc.Type == HandlerType::RESULT) {
-      ErrParam = SuccessParams.back();
-    } else if (HandlerDesc.HasError) {
-      assert(HandlerDesc.Type == HandlerType::PARAMS);
-      ErrParam = SuccessParams.back();
-      SuccessParams = SuccessParams.drop_back();
-    }
-
+    ClosureCallbackParams CallbackParams(HandlerDesc, Callback);
     ClassifiedBlocks Blocks;
+    auto *CallbackBody = Callback->getBody();
     if (!HandlerDesc.HasError) {
       Blocks.SuccessBlock.addNodesInBraceStmt(CallbackBody);
     } else if (!CallbackBody->getElements().empty()) {
-      llvm::DenseSet<const Decl *> UnwrapParams;
-      for (auto *Param : SuccessParams) {
-        if (HandlerDesc.shouldUnwrap(Param->getType()))
-          UnwrapParams.insert(Param);
-      }
-      if (ErrParam)
-        UnwrapParams.insert(ErrParam);
-      CallbackClassifier::classifyInto(
-          Blocks, HandlerDesc.Func, SuccessParams, HandledSwitches, DiagEngine,
-          UnwrapParams, ErrParam, HandlerDesc.Type, CallbackBody);
+      CallbackClassifier::classifyInto(Blocks, CallbackParams, HandledSwitches,
+                                       DiagEngine, CallbackBody);
     }
 
+    SmallVector<const ParamDecl *, 4> Scratch;
+    auto SuccessBindings = CallbackParams.getSuccessParamsToBind(Scratch);
```

Comment:
Given this is the only use I'd prefer just returning the SmallVector from `getSuccessParamsToBind`.

And not sure if this part is worth it or not but...
  - Could name `SuccessBindings` just `Bindings`
  - Add an `ArrayRef` `SuccessBindings` of `Bindings`
  - Remove `AllBindings` and just add `ErrParam` to `Bindings` in the if

Reply:
Hmm, returning a SmallVector directly would mean the callee would have to specify the inline buffer size, which feels odd – maybe we could return a `llvm::TinyPtrVector` or `std::vector`?

Regarding mutating a single `Bindings` vector directly, I tend to prefer reducing mutability where I can, but this is all pretty localised so I don't feel too strongly about it

Reply:
I ended up generalising this logic to not need the mutation. I did switch over using `llvm::TinyPtrVector` though, I don't really feel too strongly about the return type tho

Reply:
`TinyPtrVector` is optimised for 0/1 right? I guess that is probably almost always true for the success/error. I'm fine with specifying the inline buffer size in the return type, but I don't *really* mind. Personally I'd still use `SmallVector<Thing, 0>` over a vector, even if I was returning it.

Reply:
Yeah `TinyPtrVector` is optimized for the single element (and empty) case, which as you say is fairly likely for the success/error params. Out of curiosity, what benefits would `SmallVector<Thing, 0>` have over a regular vector?